### PR TITLE
Flows - Affine and Masked Affine

### PR DESCRIPTION
**Pull request by shreyasvinaya**
## Description

This PR adds an abstract class for Flow based models, 
Affine Flow is now added to flows, 
Masked affine flow is a new layer that is added

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/flows.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
+""" This module contains the implementation of the various flow layers and models"""
+import numpy as np
+import torch
+import torch.nn as nn
+from typing import Optional, Sequence, Tuple, Union
+
+
+class Flow(nn.Module):
+    """
+    Generic class for flow functions
```
Comment (rbharath) [inline deepchem/models/torch_models/flows.py:10]: More docs. Add at least a paragraph with references

File: deepchem/models/torch_models/flows.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+        -------
+        z_: torch.Tensor
+            Transformed tensor
+        log_det: torch.Tensor
+            Logarithm of the determinant of the Jacobian of the transformation
+        """
+        raise NotImplementedError("This flow has no algebraic inverse.")
+
+
+# Adding duplicate class here, the layers `Affine` class to be deprecated
```
Comment (rbharath) [inline deepchem/models/torch_models/flows.py:55]: Why are we deprecating that class? We should modify or extend it instead or name this class something different

**Review by rbharath — COMMENTED** [review]
Code is fine once duplication issue is fixed and tests pass

File: deepchem/models/torch_models/flows.py (hunk 2 above)
Comment (shreyasvinaya) [reply deepchem/models/torch_models/flows.py:55]: done

File: deepchem/models/torch_models/flows.py (hunk 1 above)
Comment (shreyasvinaya) [reply deepchem/models/torch_models/flows.py:10]: Done

**Review by rbharath — COMMENTED** [review]
LGTM once CI is clear

**Review by rbharath — APPROVED** [review]
LGTM
